### PR TITLE
add support for all Arrow IPC formats in roapi-http

### LINKED ISSUE #66: Support loading of Arrow IPC format
I think it would be useful if ROAPI can not only use Arrow IPC for response encoding, but also as a dataset format.

### PR DESCRIPTION
**Pull request by ekroon**
Implementation for #66 

Outstanding questions
- This implements only support for Arrow IPC stream, because ROAPI doesn't support Arrow IPC file output. I am not sure this rational holds...
- Would it make sense to add an integration test which tests from CLI to HTTP api? There is a lot of wiring code that is not tested at the moment (apart from the missing unit test in `arrow_stream.rs` which I will still add).

Closes #66

### CODE REVIEW

File: columnq/src/table/arrow_stream.rs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+use log::debug;
+
+use crate::error::ColumnQError;
+use crate::table::TableSource;
+
+pub async fn to_mem_table(
+    t: &TableSource,
+) -> Result<datafusion::datasource::MemTable, ColumnQError> {
+    debug!("inferring arrow table schema...");
+    let schema_ref: arrow::datatypes::SchemaRef = match &t.schema {
```
Comment (ekroon) [inline columnq/src/table/arrow_stream.rs:15]: schema and partitions can be retrieved in 1 step

File: columnq/src/table/mod.rs [hunk 2]
```diff
@@ -140,6 +141,7 @@ pub enum TableLoadOption {
     parquet {},
     google_spreadsheet(TableOptionGoogleSpreasheet),
     delta {},
+    arrow_stream {},
```
Comment (houqp) [inline columnq/src/table/mod.rs:144]: how about we keep the name as `arrows` to be consistent with other file formats?

File: columnq/src/table/arrow_stream.rs [hunk 3]
```diff
@@ -0,0 +1,48 @@
+use std::sync::Arc;
```
Comment (houqp) [inline columnq/src/table/arrow_stream.rs:1]: would be good if we name this module arrow_ipc_stream

**Review by houqp — COMMENTED** [review]
Thanks @ekroon ! the implementation looks good to me :+1: 

> This implements only support for Arrow IPC stream, because ROAPI doesn't support Arrow IPC file output. I am not sure this rational holds...

The output format (encoding) and input table format is fully decoupled. For example, we don't support google spreadsheet as output format, but it's a valid input table format :) So it's totally fine for us to add Arrow IPC file format as well if you want to add it.

> Would it make sense to add an integration test which tests from CLI to HTTP api? There is a lot of wiring code that is not tested at the moment (apart from the missing unit test in arrow_stream.rs which I will still add).

More tests are always welcome :)

**Comment by ekroon** [comment]
@houqp I decided to add support for Arrow IPC File encoding too and this way allow for all the combinations via HTTP, let me know what you think about it.

I would like to add new issues for integration tests and adding support for all the formats in `columnq-cli`, do you agree?

**Comment by ekroon** [comment]
Documentation changes are still work in progress :)

File: columnq/src/table/arrow_ipc_stream.rs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+use crate::table::TableSource;
+
+pub async fn to_mem_table(
+    t: &TableSource,
+) -> Result<datafusion::datasource::MemTable, ColumnQError> {
+    debug!("inferring arrow table schema...");
+    let schema_ref: arrow::datatypes::SchemaRef = match &t.schema {
+        Some(s) => Arc::new(s.into()),
+        None => {
+            let schemas = partitions_from_table_source!(t, |mut r| {
```
Comment (houqp) [inline columnq/src/table/arrow_ipc_stream.rs:18]: same here, we can get the schema from the partitions below to avoid extra round of IO.

**Review by houqp — APPROVED** [review]
The rest looks good to me, thank you @ekroon !

**Comment by ekroon** [comment]
@houqp Should be better now. I think there is quite some code duplications that can be reduced quite a lot, but I am out of time for now :) Feature and performance wise it is good to merge nevertheless.

**Review by houqp — APPROVED** [review]

**Comment by houqp** [comment]
Merged, thank you @ekroon for the new features :)

**Pull request merged**